### PR TITLE
MANOPD-78395 Haproxy maintenance mode improvements

### LINKED ISSUE #250: kubeapi might not work if VIP with maintenance-type: "not bind" is specified first
When we calculate `control_plain.internal` during [enrichment](https://github.com/Netcracker/KubeMarine/blob/0.8.0/kubemarine/core/defaults.py#L291) we do not check if VIP has maintenance-type: "not bind". `control_plain.internal` might become that IP if it is specified first in the list. For example,
```yaml
vrrp_ips:
- ip: 192.168.0.1
  floating_ip: 10.101.0.1
  params:
    maintenance-type: "not bind"
- ip: 192.168.0.2
  floating_ip: 10.101.0.2
```

Result:
```
Control plains:
    Internal: 192.168.0.1 (vrrp_ip[0])
```

It needs to check what will happen with internal kubeapi traffic in maintenance mode and probably fix enrichment procedure correspondingly.

### PR DESCRIPTION
**Pull request by ilia1243**
### Description
* `kubectl` in general and `check_paas` procedure in particular do not work when cluster has haproxy in maintenance mode and if VRRP IP with maintenance-type: "not bind" is specified first in the inventory. The reason is because
    * haproxy in maintenance mode listens all VRRP IPs except "not bind".
    * The first "not bind" VRRP IP is chosen as `control_plain.internal` which is registered in /etc/hosts and coredns.
    * kubectl goes through `control_plain.internal` which is not listened by the balancer.
* Balancer in maintenance mode does not listen its own internal_address, so it is impossible to send requests directly to it.
* There is redundant requirement for non-HA clusters to have at least two VRRP IPs configured for haproxy with maintenance mode. In fact, only one VRRP IP that is with `maintenance-type: "not bind"` is enough.

Fixes #250 

### Solution
* #216 is partially reverted (changes in check_paas).
* VRRP IP with maintenance-type: "not bind" is now skipped when calculating of `control_plain`.
* Haproxy in maintenance mode not also listens the node's own internal address, if the balancer is not combined with other roles.
* Now only one VRRP IP (with `maintenance-type: "not bind"`) is enough for non-HA schemes with haproxy in maintenance mode.
* Added patch which migrates the cluster from not bind VRRP IP.
* Added patch which assigns `internal_address` for haproxy in maintenance mode for existing clusters.
* Added ability to deviate enrichment of cluster to simplify patch development.
* Removed unnecessary template haproxy_mntc.cfg.j2
* Added more robust validation of haproxy and VRRP configuration.
* Added unit tests for `control_plain` detection and for allowed/prohibited haproxy schemes.
* Other minor fixes, refactoring and changes in documentation.

### How to apply
1. Run `kubemarine migrate_kubemarine --force-apply migrate_not_bind_vrrp`

**Note**: the migration changes [control_plain](https://github.com/Netcracker/KubeMarine/blob/main/documentation/Installation.md#control_plain) if it was previously chosen as VRRP IP with `maintenance-type: "not bind"` to the other best suitable one by the algorithm. You may prefer another way to resolve the problem, for example by changing the VRRP IPs that should be `not bind` and reconfiguring of haproxy. If you choose to use `migrate_kubemarine`, please still consider other custom usages of `control_plain` and handle this accordingly.

2. Run `kubemarine migrate_kubemarine --force-apply haproxy_maintenance_listen_internal`

### Test Cases

**TestCase 1**

Test Configuration:

- Hardware: Any
- OS: Any
- Inventory: 1) haproxy has maintenance mode; 2) VRRP IP with maintenance-type: "not bind" is specified first in the `vrrp_ips` list.

Steps:

1. Install cluster.
2. Switch haproxy to maintenance mode
3. Run `check_paas`.

Results:

| Before | After |
| ------ | ------ |
| many checks fail with errors or warnings | all checks succeed |

**TestCase 2**

Test Configuration:

- Hardware: Any
- OS: Any
- Inventory: 1) haproxy has maintenance mode; 2) VRRP IP with maintenance-type: "not bind" is specified first in the `vrrp_ips` list.

Steps:

1. Install cluster using older version of Kubemarine.
2. Switch haproxy to maintenance mode
3. Run `check_paas`. ER: many checks fail with errors or warnings.
4. Run `kubemarine migrate_kubemarine --force-apply migrate_not_bind_vrrp` using newer version of Kubemarine.
5. Run `check_paas`. ER: all checks succeed.

**TestCase 3**

Test Configuration:

- Hardware: Any
- OS: Any
- Inventory: full-HA with haproxy in maintenance mode

Steps:

1. Install cluster.
2. Switch haproxy to maintenance mode.
3. Send request directly to any balancer.

Results:

| Before | After |
| ------ | ------ |
| Connection refused | 404 not found or response from ingress if configured |

**TestCase 4**

Test Configuration:

- Hardware: Any
- OS: Any
- Inventory: full-HA with haproxy in maintenance mode

Steps:

1. Install cluster using older version of Kubemarine.
2. Switch haproxy to maintenance mode
3. Send request directly to any balancer. ER: Connection refused
4. Run `kubemarine migrate_kubemarine --force-apply haproxy_maintenance_listen_internal` using newer version of Kubemarine.
5. Send request directly to any balancer. ER: 404 not found or response from ingress if configured

**TestCase 5**

Test Configuration:

- Hardware: Any
- OS: Any
- Inventory: 1) non-HA with 1 control-plane/worker and 1 balancer; 2) haproxy has maintenance mode and there is only one VRRP IP that is with maintenance-type: "not bind".

1. Install cluster.
2. Switch haproxy to maintenance mode
3. Run `check_paas`.

Results:

| Before | After |
| ------ | ------ |
| Error "Needs at least one VRRP IP without 'maintenance-type: not bind'" | Cluster is installed and all checks are passed |

### Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Integration CI passed
- [x] Unit tests. If Yes list of new/changed tests with brief description
- [x] There is no merge conflicts

#### Unit tests
Added 2 tests in `test/unit/test_defaults.py` to verify changed algorithm of `control_plain` calculation.
Added and fixed unit tests in `test/unit/test_haproxy.py` for allowed and prohibited haproxy schemes, taking into account maintenance mode.

### CODE REVIEW

File: kubemarine/haproxy.py [hunk 1]
```diff
@@ -68,11 +82,24 @@ def enrich_inventory(inventory, cluster):
...
+    return exe.get_merged_result()
+
+
 def install(group):
     with RemoteExecutor(group.cluster) as exe:
         for node in group.get_ordered_members_list(provide_node_configs=True):
-            package_associations = group.cluster.get_associations_for_node(node['connect_to'])['haproxy']
-            group.sudo("%s -v" % package_associations['executable_name'], warn=True)
+            package_associations = get_associations_for_node(node)
+            node['connection'].sudo("%s -v" % package_associations['executable_name'], warn=True)
```
Comment (ilia1243) [inline kubemarine/haproxy.py:102]: Note that this is not only a refactoring but also some small fix

File: kubemarine/haproxy.py [hunk 2]
```diff
@@ -68,11 +82,24 @@ def enrich_inventory(inventory, cluster):
...
 
 
+def get_config_path(group: NodeGroup) -> NodeGroupResult:
+    with RemoteExecutor(group.cluster) as exe:
+        for node in group.get_ordered_members_list(provide_node_configs=True):
+            package_associations = get_associations_for_node(node)
+            cmd = f"systemctl show -p MainPID {package_associations['service_name']} " \
+                  f"| cut -d '=' -f2 " \
+                  f"| xargs -I PID sudo cat /proc/PID/environ " \
+                  f"| tr '\\0' '\\n' | grep CONFIG | cut -d \"=\" -f2 | tr -d '\\n'"
```
Comment (ilia1243) [inline kubemarine/haproxy.py:92]: Use the same way to detect currently used config as external-traffic-manager does.

**Review by shmo1218 — APPROVED** [review]
The doc file is reviewed.

File: documentation/Installation.md [hunk 3]
```diff
@@ -2623,10 +2623,10 @@ The following settings are supported:
     <td>By enabling ready, an HTTP endpoint on port 8181 returns 200 OK when all plugins that are able to signal readiness have done so. If some are not ready, the endpoint still returns a 503 with the body containing the list of plugins that are not ready. Once a plugin has signaled that it is ready, it is not queried again.</td>
   </tr>
   <tr>
-    <td>prometheus</td>
+    <td>Prometheus</td>
```
Comment (ilia1243) [inline documentation/Installation.md:2626]: Hi @shmo1218 
This is a key. It might be case sensitive.

Comment (shmo1218) [reply documentation/Installation.md]: Hi, @ilia1243,

Thank you for pointing it out. I have rollbacked the change.

**Review by theboringstuff — APPROVED** [review]

**Review by theboringstuff — APPROVED** [review]

**Pull request merged**